src/AESecb.cpp

```diff
@@ -0,0 +1,139 @@
+/**
+  @file AESecb.cpp
+  Implementation of ECB mode of operation for AES-128 algorithm
+*/
+#include "AESecb.hpp"
+#include <iostream>
+
+
+void printEncryptOutput(std::vector<unsigned char>& output) {
+  std::cout << std::endl;
+  std::cout << "Encrypt result";
+  for (int i = 0; i < output.size() / 16; i++){
+    std::cout << std::endl;
+    std::cout << "Block " << i << ": ";
+    for (int j = 0; j < 16; j++) {
+      std::cout << std::hex << (int) output[j+(i*16)];
+      std::cout << " ";
+    }
+  }
+  std::cout << std::endl;
+}
+
+
+/**
+  Implements cipher with ECB mode of operation
+  @param input: vector of hex values representing plaintext
+  @param output: vector of hex values representing (padded) ciphertext
+  @param key: vector of hex values representing key to use
+  @return none
+*/
+void encrypt_ecb(const std::vector<unsigned char>& input, std::vector<unsigned char>& output, const std::vector<unsigned char>& key) {  
+  // Calculate padding length, then copy input array and padding into plaintext
+  const std::size_t inputSize = input.size();
+  const std::size_t padLength = 16 - (inputSize % 16);
+  const std::size_t plaintextLength = inputSize + padLength;
+
+  // Plaintext accomodates both the input and the necessary padding
+  std::vector<unsigned char> plaintext;
+  plaintext.reserve(plaintextLength);
+  plaintext = input;
+
+  for (int i = 0; i < padLength; i++) {
+    // PKCS#7 padding (source: https://www.ibm.com/docs/en/zos/2.1.0?topic=rules-pkcs-padding-method)
+    plaintext[inputSize+i] = padLength;
+  }
+
+  // Loop over number of blocks
+  for (int i = 0; i < plaintextLength / 16; i++) {
+
+    // Loop over block size and fill each block
+    std::array<unsigned char, 16> block;
```

Comment:
Is there an advantage to the overhead by initializing the block and outputBlock arrays before the loop? Or would it be more secure to limit their scopes?

Reply:
The block is 16 bytes and each iteration (out of 16 iterations) of the inner loop (line 52) corresponds to a byte, so initalizing the block inside the inner loop would mean blocks of 1 byte. The outer loop (line 48) is actually there for the purpose of initializing and iterating over each block.